add API for on-off service request and release management

pabigot opened the pull request:
There are various situations where it's necessary to support turning devices on or off at runtime based on demand from multiple clients.  This includes power rails, clocks, other peripherals, and binary device power management.  The complexity of properly managing multiple consumers of a device in a multithreaded system, especially when transitions may be asynchronous, suggests that a shared implementation is desirable.

This commit presents a service API with these features:

* The stable states are `OFF` (initial) and `ON`.
* The core operations are request (add a dependency) and release (remove a dependency).  The service manages the state based on these operations.
* Operations are initiated by a function call referencing a specific service and passed client notification data.  The function call will succeed or fail.  On success, the operation is guaranteed to be initiated, but whether the operation succeeds or fails is indicated by the client notification.
* Service configuration provides functions that implement transitions from off to on (start), from on to off (stop), and optionally reset from error.  Transitions may be marked as synchronous or asynchronous; this is used to prevent unintended blocking.
* Transition from OFF to ON when first client request is received;
* Transition from ON to OFF when last client release is received;
* Requests may be queued while the system is transitioning to on or off.  If queued when transitioning to off, the service will initiate a transition to on as soon as the transition to off completes.
* Requests are reference counted, but not tracked.  I.e. clients are responsible for recording whether their requests were accepted, and improper releases can cause an active requester to be shut out.
* Failures in transition are recorded and inhibit further requests or releases until the service is reset.  Pending requests are notified (and cancelled) when errors are discovered.
* Transition operations are executed in one of the following modes:
  * Spin-wait: After invoking the function the caller delays until the operation has completed.
  * Signals: A pointer to a `k_poll_signal` is a parameter, and it is raised when the transition completes.  The operation completion code is stored as the signal value.
  * Calls back: a function pointer is provided, and the function is invoked with the operation completion code.
* Synchronous transition may be implemented by a caller (e.g. via poll() until completion is signalled).

pabigot committed:
lib: os: onoff: add API for on-off service request and release management

There are various situations where it's necessary to support turning
devices on or off at runtime, includin power rails, clocks, other
peripherals, and binary device power management.  The complexity of
properly managing multiple consumers of a device in a multithreaded
system suggests that a shared implementation is desirable.  This
commit provides an API that supports managing on-off resources.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
While I personally need this for power rail control, I designed it to support device power management needs as well.  I also believe it could be used to replace the Nordic clock enable/disable infrastructure by providing all the resource management/thread-safety implementation generically, allowing the clock implementation to only deal with start and stop tasks and events.  This could reduce the need for future fixes like #20796 and would address concerns raised in its review.

pabigot commented:
Before this is ready for merge two changes need to be made to align this with #20959:
* The onoff operation API and transition functions need to be made isr-capable and documented as such;
* The `K_ONOFF_SERVICE_ASYNC_foo` flags should be renamed to `K_ONOFF_SERVICE_SUSPENDABLE_foo` to match the intent behind the API no-wait support.  The service provider functions are by design asynchronous.

mbolivar commented on include/kernel.h:
```diff
@@ -5178,6 +5178,516 @@ extern void k_mem_domain_remove_thread(k_tid_t thread);
+	 */
+	sys_slist_t clients;
+
+	/* Function to invoke to transition the service to on. */
```
Any particular reason not to put these in an API structure and use the device model?

pabigot replied:
This doesn't provide a device abstraction.  It's an implementation of a data structure where a small part of the behavior is specific to a particular service, specified by injecting functions though the service configuration data.

In many cases user code won't be looking to access the service directly.  E.g. a clock driver might have an onoff service object as part of the driver state, and would delegate its public on/off API methods to calls to the request and release operations given that service object.  There wouldn't even be any devicetree elements associated with the service.

In other cases there would be devicetree information used to generate the service object.  I expect a power rail to be described in devicetree as something like:
```
   sensor-rail: sensor-rail {
      compatible = "regulator,gpio";
      rail-gpios = <&gpio1 8 GPIO_ACTIVE_LOW>;
      on-delay = <500>; /* ns */
   };

   sht3xd@3c {
      /* ... */
      power-rail = &sensor-rail;
   };
```
Ideally this would produce a service object definition that could be referenced from the sensor driver's device power management functions, which would be populated with references to generic gpio_regulator_onoff_{start,stop} functions.  Even there there's no real motivation to make the rail a device; we just need some way for the sensor driver to get a reference to the service object.

For that to be done automatically there will have to be devicetree scripting enhancements.  It may well turn out that to be practical in the near term there'd have to be a device associated with the `gpio,regulator` compatible that would implement a generic "regulator" API.

I hope to prototype something like this within the topic-gpio branch in the next few days.

mbolivar replied:
Thanks for the response! It seemed like it was probably somewhat orthogonal, but it's good to have a clear answer like this for the record.

mbolivar reviewed (commented):
From a high level the approach and API seem reasonable and well-defined. I won't review the details until we've nailed down some more terminology. One general question.

pabigot commented:
Based on external feedback this is now in `lib/os` parallel to features like `ring_buffer`.

nordic-krch commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,533 @@
+ * @retval -EWOULDBLOCK if a non-blocking request was made and could
+ *         not be satisfied
+ */
+__syscall int onoff_request(struct onoff_service *srv,
```
When "signal type" client needs to call `onoff_service_has_error()` after `onoff_request()`? Only when `-EIO` is returned?

pabigot replied:
When `onoff_request()` returns `-EO` it means the service already had an error at the point the call was made; there's no need to check has_error because it will return true (unless something slips in and resets).

If the request produces an error, that's indicated in the operation result (for no-wait, signal, or callback).  There the negative result indicates a failure to complete the transition, and any error puts the service into an error state (because it can't tell if the failure left the system in the original state or an undefined state).

nordic-krch reviewed (commented):
Comments to `include/sys/onoff.h`. Will continue review of other parts.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,503 @@
+
+		break;
+	case SERVICE_STATE_TO_ON:
+		rv = -EBUSY;
```
this can be quite error prone. Let say that module A started the clock but before it got started clock is no longer needed. Module A releases the clock but gets `-EBUSY`. Now, there must be a logic in the notification callback which on successful clock starting needs to check the status which must say something like: disable but failed to release the clock. Based on that status release must be called again.

An alternative would be to decrement the counter, remove client from list and return success. Then in `onoff_start_notify` trigger stop if `ref_cnt == 0`. Return `-EBUSY` only if client is not found in the list. That would simplify early service releasing.

pabigot replied:
Requesting a start, then requesting a stop before the start completes, is functionally a "cancel in-progress operation" request.  The API currently doesn't support that, but there are reasons why it should, and I'll take a look at that later.  It would be a separate function.

nordic-krch replied:
@pabigot do you plan to add this feature in this PR or as next step once this get merged?

pabigot replied:
It's already implemented.  `onoff_cancel()`.

nordic-krch reviewed (commented):
Bunch of comments for `onoff.c`

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,533 @@
+	 * When provided this indicates the start transition function
+	 * may suspend.
+	 */
+	ONOFF_SERVICE_SUSPENDABLE_START         = BIT(1),
```
The intent of these flags is to specify transition function behavior as discussed in #20959, which is being actively revised; `suspend` is currently `rescheduling` or maybe `cooperative-safe`.

The transition functions may be invoked from thread or interrupt context.  At the moment there's a distinction between `isr-callable` and what is now `reschduling`.  The service can know whether it's being run from an ISR or a thread.  The flags are used to tell the service what the behavior of the transition function is, since it can't look at documentation, and from that determine a request can be satisfied.

At the moment there is only one code path that differs in behavior between the features, and it probably shouldn't.

This may get resolved when we settle on the terms.  It would be convenient if there was only one flag.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,533 @@
+ * ONOFF_SERVICE_ISR_CALLABLE_STOP and/or
+ * ONOFF_SERVICE_SUSPENDABLE_STOP as appropriate in flags.
+ *
+ * @param reset the function used to clear errors and force the
```
Service initialization is the responsibility of whatever calls `onoff_service_init()`: that must make sure the state of the controlled thing is off.  However that function must not be invoked by clients, so there should be some sort of information hiding added.

See more below about reset.

nordic-krch replied:
> Service initialization is the responsibility of whatever calls onoff_service_init()

Then it should be added to the function description

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,533 @@
+ * the client.
+ */
+enum onoff_client_flags {
+	/* Indicates operations are to proceed only if the initiating
```
It's intended to mean what `K_NO_WAIT` means for time-limited operations: if you can do it immediately fine, if not, fail.  This too needs to be updated for terminology changes in #20959.

nordic-krch replied:
do you see a use case of such mode?

pabigot replied:
Absolutely.  Many things that can be turned on and off will never delay (part of why "suspendable" is a service flag).  All of them will only delay on the first or last request.  If the caller does not expect a delay to be required the caller can allocate the client structure on the stack and be guaranteed a reference to it will not be held in the service.

It's this feature that makes it possible to request and release resources from ISRs, even if you still need to provide a way to re-issue the operation from thread context if it fails.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,533 @@
+	 * async infrastructure.  This field is the sole record of
+	 * success or failure for no-wait synchronous transitions.
+	 */
+	int result;
```
Possibly.  It's useful for callback because the callback handler doesn't have to store the result somewhere for follow-on processing.  I'll look into moving it.

nordic-krch replied:
may a service should store the result. It should be common for all clients since it comes from `onoff_service_notify_fn`

pabigot replied:
It can't be held by the service because it could be overwritten at any time by another transition initiated by another thread/ISR.  Clients that were delayed in looking for it would get the wrong result.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,533 @@
+			    struct onoff_client *cli);
+
+/**
+ * @brief Release a reserved use of an on-off service.
```
"Release" means that all accounting related to the specific client's release has been completed.  It does not mean that the service has been turned off, unless that client was the last one active.  Only the last client to release will ever get a delayed result, and then only if the stop function delays.  I'll update the documentation to make this clear.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,533 @@
+	sys_snode_t node;
+
+	union {
+		/* Pointer to signal used to notify client.
```
Whatever operation is currently active using this structure: request, release, or reset.

nordic-krch replied:
haven't you considered having operation type argument (req, rel, reset)? This is known to the service when notifying client. Client need to have additional service state to handle that properly.
e.g.
```
void notification(struct onoff_service *srv,
				      struct onoff_client *cli,
				      int res)
{
  client_owner = container_of(cli);
  if ((client_state== SRV_REQUESTED) && (res < 0)) {
     client_state = RESETTING;
    onoff_service_reset(...);
  } else {
  ...
  }
}
```
As you can see, this handling gets complicated.

I'm also missing `user_data` in that callback. Need to use `container_of` and it is not always the case that it can be used to get entity owning the client.

pabigot replied:
No, I hadn't, because the client provides the structure initialized for the specific operation.  I would expect something like:

```
static void notified_requested(struct onoff_service *srv,
                               struct onoff_client *cli,
                               int res)
{
  if (res < 0) {
     cli->async.callback = notify_reset;
     int rc = onoff_service_reset(srv, cli);
     if (rc < 0) {
        do_something();
     }
  } else {
     do_something_else();
  }
}
```
I think the reaction to each specific transition type is likely to be completely distinct.  Until there's evidence that a common callback provides significant value I'd rather leave it as is.

I'll add `user_data`, though in all cases I've come up with so far `CONTAINER_OF()` is sufficient.  It would be nice if we could make that a standard feature of all asynchronous callback APIs, but when I tried a year ago I got significant push-back.

nordic-krch replied:
> I'll add user_data, though in all cases I've come up with so far CONTAINER_OF() is sufficient.

It is not sufficient if client owner is have pointer to client:
```
struct client cli;
struct owner {
  struct client *cli;
}

struct client cli;
const struct owner o = {.cli = &cli};
```

Regarding changing the callback. You can of course do that but that requires callback element to be in ram. However, in general i prefer to keep separate constant parts of the module from volatile parts to reduce ram usage (`struct shell` as the example) and even this is not the case now in client structure i would favor parameter in the callback.

pabigot replied:
The `onoff_client` structure that holds the notification details must be in RAM, because it has to be linkable into a list.  So you'll always be able to change the callback (my example code was "wrong" because it did it directly; it should have used `onoff_client_init_callback()`.)

nordic-krch replied:
i was not suggesting that it can be done here. I was rather pointing out that this approach completely closes this option.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,533 @@
+
+/**
+ * @brief Signature used to notify an on-off service client of the
+ * completion of a transition.
```
Reset is a transition from an error state to the off state: there is no distinction.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,533 @@
+__syscall bool onoff_service_has_error(const struct onoff_service *srv);
+
+/**
+ * @brief Clear errors on an on-off service and reset it to its off
```
It doesn't.  They get queued just like any other transition request, and the client will be notified when the reset is complete.  If a client tries to reset after complete the call will fail, because reset is rejected if the service is not in an error state.  That's subtly indicated by one of the `EINVAL` retval descriptions but should be more clear: I've queued some rewording.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,533 @@
+	/* List of clients waiting for completion of reset or
+	 * transition to on.
+	 */
+	sys_slist_t clients;
```
No, all clients in the list at any time are guaranteed to be waiting for the same transition.  As soon as a transition completes a "copy" of the client list is made and the one in the service state is cleared.  Clients are then notified from the copy.

nordic-krch replied:
ok, got it. Didn't notice that there is a list copy created.

pabigot commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,503 @@
+	srv->flags |= (state & SERVICE_STATE_MASK);
+}
+
+static int validate_args(const struct onoff_service *srv,
```
It checks whether `srv` is null.

pabigot commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,503 @@
+		set_service_state(srv, SERVICE_STATE_OFF);
+	} else if (k_is_in_isr()
+		   && ((srv->flags
+			& ONOFF_SERVICE_ISR_CALLABLE_START)
```
No; this is checking whether it's safe to restart the service if somebody queued a request while the stop was pending.

pabigot commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,503 @@
+	 * pending requests; otherwise if there are pending requests
+	 * start the transition to ON.
+	 */
+	notify_one(srv, releaser, res);
```
I think I see some source of confusion here.  Each transition from a client passes client state, which is used solely to provide notification of completion of that transition.  The service does not retain that state, and the notification will occur exactly once when that operation completes, at which point ownership of the state reverts to the client.

This is what was meant in the description above:

> Requests are reference counted, but not tracked. I.e. clients are responsible for recording whether their requests were accepted, and improper releases can cause an active requester to be shut out.

An alternative solution would have a permanently associated state held by the service as long as the client request was active, but that ends up making the solution much harder.

pabigot commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,503 @@
+		}
+
+		/* Reject if in isr but stop not isr-callable */
+		if (k_is_in_isr()
```
Yes.  ISRs are not excluded from the requirement to check for potential failures and be prepared to handle them.  If the final stop causes a suspend, the caller must be prepared to resubmit the attempt from thread context.  However that should not prevent the operation from completing in cases where it will not suspend (all that's being done is decrementing a reference count).

pabigot commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,503 @@
+		/* Queue to start after release */
+		__ASSERT_NO_MSG(srv->releaser != NULL);
+		add_client = true;
+		rv = 3;
```
No.  These are not part of the API, they're used solely for testing, and there's no place to share private information between the implementation and the test.

pabigot commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,503 @@
+		reset = true;
+		srv->flags |= SERVICE_STATE_TRANSITION;
+	}
+	sys_slist_append(&srv->clients, &cli->node);
```
This was answered above: the client list is managed atomically so that when new clients come in they're only appended if they're asking for the currently pending operation, and all clients are notified as soon as the operation completes, and after the list has been reset for the next operation.

nordic-krch commented:
I've been thinking that service could be extended with observers. Observer does not request or release service but is notified when state is changed. A particular use case is clock calibration which is started when LF clock is started and stopped when LF clock is stopped.

pabigot commented:
It could, but I think that's out of scope.  The particular case you're interested in can easily be handled: the code that starts and stops the clock can also start and stop calibration, either directly or using another onoff service.

I have draft design for a more powerful resource coordination infrastructure which, unlike this one, would be "tracked" (all clients are notified of all potential and actual state changes) from which observer capability just falls out.  Details to follow.

pabigot commented:
Latest update:

* Switched back to a single flag (`FOO_WAITS`) to validate use of API in no-wait and ISR contexts, because it seems "not no-wait" is the condition for ISR callable;
* Added `user_data` to callback API;
* Implemented `onoff_cancel()` to wrest back control of client state structures when a service operation is pending;
* Added documentation page;
* Added coverage for all non-trivial branches

I don't anticipate any further changes in functionality or naming at this point.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,542 @@
+		__ASSERT_NO_MSG(srv->refs == 0);
+		set_service_state(srv, SERVICE_STATE_TO_ON);
+		start = true;
+		add_client = true;
```
i think that this one is not right. In case `no wait` client is triggering start (first requestor) then it should not be added to the list. Should be something like `add_client = !no_wait`.

pabigot replied:
No, this is right.  It's already been checked that the start call will complete without waiting, and part of the start is invoking `onoff_notify_start()` which notifies everything in the client list.

The only time you don't record the client is when the service is already in the desired state so the transition function isn't invoked.

nordic-krch replied:
but if `no wait` client is created on stack it cannot be enqueued, it just wants to trigger start and increase ref counter

nordic-krch replied:
in case of clock driver, sys timer wants to request LF clock and it does not care when it is started. cannot use signal because it's before kernel and callback would require to keep the client variable. i want to use stack-allocated, no-wait client there.

pabigot replied:
It can be enqueued, because the start function will complete, and the list will be processed and cleared, before the `onoff_service_request()` call returns.  Completion of the notification is what transfers control of the client state back to the application at which point the structure can be cleared.

nordic-krch replied:
> It can be enqueued, because the start function will complete, and the list will be processed and cleared, before the onoff_service_request() call returns.

No, that's not true. If it is the first request then start function will trigger clock starting which will complete in hundreds of milliseconds. Since it was in `SERVICE_STATE_OFF` it has `add_client=true` and `start=true`. Client is put into the list and start function is called and function leaves. ref is not incremented. In start notify (called from clock isr) list is processed but client on that list is no longer valid because it was allocated on the stack.

pabigot replied:
There's some confusion about no-wait.  No-wait means "do not execute this unless the entire operation will be complete at the time the function returns."  You cannot successfully request a no-wait start when the transition will wait.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,542 @@
+out:
+	if (add_client) {
+		sys_slist_append(&srv->clients, &cli->node);
+	} else if (notify) {
```
`else if (notify || !no_wait)`?
In case of `no wait` client counter should also be incremented.

pabigot replied:
That will be done in `onoff_notify_start()` because the no-wait client is in the list.

pabigot commented:
To recap and complete off-line discussion started from https://github.com/zephyrproject-rtos/zephyr/pull/21090#discussion_r355404802:

You cannot do a no-wait request for a waiting operation because the API does not allow you to discard the result of the operation (whether it succeeds or fails), and if you don't have a client state structure recorded there's no way to notify the client.

This API is unconditionally asynchronous, but as noted synchronous transition may be implemented by a caller.  While the original description above only mentioned `k_poll()`, this API must be usable from pre-kernel stages where that's not available.  You do that by using a stack-based client state with a callback notification where the caller waits for the operation to complete before it returns:

```
static void clock_started(struct onoff_service *srv,
                          struct onoff_client *cli,
			  void *user_data,
			  res)
{
    __ASSERT_NO_MSG(res >= 0);
    *(volatile bool *)user_data = true;
}

static int custom_clock_control_on(struct device *dev,
                                   clock_control_subsys_t sys)
{
   struct custom_data *data = (struct custom_data *)dev->driver_data;
   struct onoff_client cli;
   volatile bool started = false;

   onoff_client_init_callback(&cli, clock_started, (void*)&started);
   int rc = onoff_service_request(&data->srv, &cli);
   if (rc >= 0) {
      while (!started) {
         /* spin */
      }
      rc = cli.result;
   }
   /* error check here */
   return rc;
}
```

I will be adding something like this to a sample and incorporating it into the documentation before this PR gets out of draft state.

nordic-krch commented:
> You cannot do a no-wait request for a waiting operation because the API does not allow you to discard the result of the operation

What if starting service is guaranteed (or sw assumes that it is guaranteed) to be successful (like triggering clock start is expected to result in clock being started)? Currently no-wait client is capable only of subscribing to the service and not to trigger start. It means that it behavior depends on other clients. Also right now, using no-wait client to start service returns no error.

By adding two proposed changes, no-wait clients can start service. Also if service is already started then result field in client structure will be filled with 0. What if `result` field is filled initially with `-EINPROGRESS` then if no-wait client initiate service and it is not completed before leaving then client is not enqueued and result is `-EINPROGRESS`, if service is already started then result is set to 0.

pabigot commented:
> Currently no-wait client is capable only of subscribing to the service and not to trigger start. It means that it behavior depends on other clients.

That isn't true.  You can issue a no-wait start as long as the start transition on does not wait.

>  Also right now, using no-wait client to start service returns no error.

Only if the start transition does not wait.  If it would wait, it returns `-EWOULDBLOCK` as it should.  The success or failure of the operation is distinct from the success or failure of the initiating function, even for no-wait, so it has to be returned through the client state notification.

Your proposed change breaks the architectural model that operation completion results are *always* provided (which the application should check, per upcoming coding guidelines).    I do not want no-wait clients to be able to do what you propose: it adds confusing special-case behavior.

> By adding two proposed changes, no-wait clients can start service.

Beyond violating the architectural principles, this would also break other capabilities related to queuing operations (like restart while a stop is pending) and cancelling operations.  Having a ref increment in two places, one of them done before you know whether the start transition will be successful, makes it much harder if not impossible to correctly manage the system.  I had originally implemented no-wait by passing a null client structure pointer.  I had to remove that, because it prevented support for more important capabilities (specifically start-after-stop: it is very hard for an application to recover if a request is rejected due to an incomplete stop from the last release).

We really do need to retain the rule that the client state associated with an incomplete operation remains owned by the driver until the operation completes, even if that means a caller has to wait so it can safely discard the state object on the stack.

nordic-krch commented:
Actually, what i would like to get is anonymous requests (with `NULL` client). Given that there is a service where start operation will succeed (like hw clock start) or there is a user which does not care about result of start operation then anonymous requesting is welcomed. There could be a service flag telling if anonymous requests are accepted. Anonymous requests can be tracked using independent counter then it shouldn't be difficult to manage the service.

Without anonymous requests, I will need to statically allocate callback client just to one time start the clock and waste 20 bytes of RAM for client structure and implement empty callback which i really don't need.

nordic-krch commented:
Also note that clock control API (`clock_control_async_on`) allows anonymous requests (`struct clock_control_async_data *data` set to NULL). Without anonymous requests it will be hard to provide backward compatibility layer.

pabigot commented:
@nordic-krch OK, I've pushed a commit that gives you an anonymous request capability.

Fair warning: when it comes time to discuss this in the context of resource management APIs, I will argue that this is a very bad idea and the capability should be removed.

The clock control async API should not be allowing you to start a clock and assume that the start was successful, nor to allow the clock to be used when it has not been fully started.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,627 @@
+	srv->flags |= (state & SERVICE_STATE_MASK);
+}
+
+/* Returns the client notification mode */
```
or negative error code.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,627 @@
+		srv->flags = set_anonymous_count(srv->flags, 0U);
+
+		/* Mark an error if reference count would overflow. */
+		if (srv->refs > (SERVICE_REFS_MAX - refs)) {
```
it is quite unlikely that u16 will be reached. Maybe it would be better to just add `__ASSERT_NO_MSG(srv->refs < MAX_CLIENTS)` to `SYS_SLIST_FOR_EACH_NODE`. Number of anonymous count is controlled in request function.

pabigot replied:
No, at this point I'm leaving open the possibility of reducing the size to `u8_t` and that could be overflowed.

nordic-krch commented:
@pabigot I was wondering if it would be possible to somehow extend the module to support "identified" requests. Meaning that when client requests service and service is enabled then such client ends up in the list of active clients (if certain flag is set). Service is not shut down unless there is any clients in the active list. If ref count reaches 0 and list is not empty that indicates some kind of service misuse. There would be a need to have `force off` functionality along with that.

I'm not 100% convinced if that's needed, wondering what are your thoughts.

pabigot commented:
> @pabigot I was wondering if it would be possible to somehow extend the module to support "identified" requests.

That's a different service, tentatively called "mediated" resource management, where registered clients get to contribute to deciding what the service operating level should be.  The same service will naturally provide notification of state changes.  This would be used for system level power management mediation, and possibly some device power management where applications need state change notification or finer control than just on-or-off.

At the moment I expect this PR to be closed in draft status and replaced by two new PRs derived from it: the first generalizes the notion of asynchronous results (basically pull `onoff_client` out and make it usable in any (new) asynchronous API, rather than forcing either a callback or a signal), second provides resource management services: onoff (as implemented here) and, if it turns out to be worth doing, mediated.

I expect that to be done later this month.  Before that I do need evidence this will work for power rail support, which is getting hampered by poor device management infrastructure.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,627 @@
+	rv = -EALREADY;
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if (sys_slist_find_and_remove(&srv->clients, &cli->node)) {
```
any reason why this cannot be called here https://github.com/zephyrproject-rtos/zephyr/blob/577f2ac8374a5bf6386345d534c1664ca18838b3/lib/os/onoff.c#L472 ?

pabigot replied:
If the service has not started, and the client is attempting to release it, the client is supposed to cancel the request instead.

This is the sort of slippery slope that gets started when you allow anonymous requests.  Between this, and the request to forget about a releaser, I'm going to remove that "feature": it's too hard to be sure all cases are handled.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,627 @@
+		stop = true;
+
+		set_service_state(srv, SERVICE_STATE_TO_OFF);
+		srv->releaser = cli;
```
can you add a flag to client which would prevent being assigned as a releaser? In case client is not interested in getting notified that he is a releaser.

pabigot replied:
The state of this field also affects management of the reference count.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,627 @@
+		return rv;
+	}
+
+	if (rv == ONOFF_CLIENT_NOTIFY_ANONYMOUS) {
```
probably `ONOFF_CLIENT_NOTIFY_SIGNAL` is also not supported. If we cancel then thread will stay pending forever.

pabigot replied:
No, this is right.  Whatever invokes cancel needs to deal with the signal.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,627 @@
+	} else if (k_is_in_isr()
+		   && ((srv->flags & ONOFF_SERVICE_START_WAITS) != 0U)) {
+		set_service_state(srv, SERVICE_STATE_OFF);
+		notify_clients = true;
```
i keep thinking about this case. It can really hit somebody hard. I wonder about optional `k_work` to which start operation would be deferred in that case. what if START_WAITS or STOP_WAITS is set then k_work could be required (although it may be to hard requirement since it can be handled on user side). On the other hand, blocking default work queue for unknown period ... that one is tricky.

pabigot replied:
"This case" being a deferred restart is rejected because it can't be invoked from the ISR?  Everybody who asked for it will get notification of the failure (`-EWOULDBLOCK`), and can request again from thread context.  I don't see any reason to have this service try to provide a solution.

nordic-krch replied:
The problem is that without looking into implementation user does not know that this can happen. `result` field in `onoff_client` structure is not documented. It should state what kind of errors it may contain apart from service specific.

nordic-krch commented:
what about getting function for getting current state of the service? Useful for logging. Same with api for getting current ref count.

pabigot commented:
> what about getting function for getting current state of the service? Useful for logging. Same with api for getting current ref count.

That initially seems useful, but it really isn't.  The state of the service can change at any time, based on interrupts adding or removing a request, so any captured state is obsolete as soon as the API call that provides it returns: the state you log is not necessarily the state when whatever event occurred that's motivating the log entry.  Logging the details of both the initiation and operation completion results should cover it.

Basically you can never be sure whether it's off, you can only be sure it's on if you have successfully completed a request, and you are responsible for knowing the status of your own requests.

pabigot commented:
Rebased and updated with a revised API more in line with the [draft API terminology](https://github.com/pabigot/zephyr/blob/api/terms/doc/api18970.md).  Changes include:

* The API is inherently asynchronous.  While some operations will complete immediately, the result of the operation is always separated from the result of the function initiating the operation.  There is no provision for forcing the initiation to behave synchronously.
* The no-wait feature is removed; instead all on-off API calls are isr-ok and pre-kernel-ok meaning they can be called from interrupts, but if the underlying transition would sleep the attempt will return an error.
* Added a "spin-wait" configuration for where the caller doesn't need/want asynchronous notification via signal or callback.
* Abstracted the process of extracting the result from the client state so that the object has no user-accessible fields.
* Removed runtime validation of object initialization, replacing it with asserts like other kernel API.
* Require client structures to be re-initialized before each use.
* Restricted cancellations to ensure that no operation completes without at least one client being notified of its result, which eliminates the possibility of a service entering an error state without notification.

mbolivar reviewed (commented):
> At the moment I expect this PR to be closed in draft status and replaced by two new PRs derived from it: the first generalizes the notion of asynchronous results (basically pull `onoff_client` out and make it usable in any (new) asynchronous API, rather than forcing either a callback or a signal), second provides resource management services: onoff (as implemented here) and, if it turns out to be worth doing, mediated.

Is that still the case? Is it worthwhile to review this further?

mbolivar commented:
> Is that still the case? Is it worthwhile to review this further?

EDIT: never mind, I see you upgraded this from draft status an hour ago while I was looking it over. So I assume your earlier comment is overcome by events and you want this merged now.

pabigot commented:
> > At the moment I expect this PR to be closed in draft status and replaced by two new PRs derived from it: the first generalizes the notion of asynchronous results (basically pull `onoff_client` out and make it usable in any (new) asynchronous API, rather than forcing either a callback or a signal), second provides resource management services: onoff (as implemented here) and, if it turns out to be worth doing, mediated.
> 
> Is that still the case? Is it worthwhile to review this further?

Yes, this is now live and prepared for real review.  Wherever my original comment you quoted is, I'd forgotten it and it's out of date. I tried to come up with a general API for asynchronous results and gave up because I didn't believe it'd ever get merged in isolation.

I do think this provides a good example of an idiom to support asynchronous operations both in thread-capable and pre-kernel situations.  It should be technically possible to start using that for other API, such as #21538, which could simplify things by only implementing the asynchronous API, with a straightforward synchronous wrapper.

mbolivar commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+ * @brief Initialize an on-off client to be used for a signal
+ * operation notification.
+ *
+ * Clients that use this initialization will by notified of the
```
```suggestion
 * Clients that use this initialization will be notified of the
```

mbolivar commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+ * @brief Initialize an on-off client to be used for a callback
+ * operation notification.
+ *
+ * Clients that use this initialization will by notified of the
```
```suggestion
 * Clients that use this initialization will be notified of the
```

mbolivar commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+ * @brief Release a reserved use of an on-off service.
+ *
+ * The return value indicates the success or failure of an attempt to
+ * initiate an operation to release of the resource.  If initiation of
```
```suggestion
 * initiate an operation to release the resource.  If initiation of
```

mbolivar commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+ *
+ * The return value indicates the success or failure of an attempt to
+ * initiate an operation to reset the resource.  If initiation of the
+ * operation succeeds the result of result of the reset operation
```
```suggestion
 * operation succeeds the result of the reset operation
```

mbolivar commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+ * actual release fails.  Always check the operation completion
+ * result.
+ *
+ * Clients will be notified with `-EWOULDBLOCK` without causing a
```
I find this hard to read.

Would this work?

```
 * Clients will be notified with `-EWOULDBLOCK` without causing a
 * service error if the attempt to release the resource would result in
 * a call to a stop transition that would sleep if the release call is
 * made from ISR or pre-kernel context.
```

mbolivar replied:
> I find this hard to read.

I still do

pabigot replied:
This one went away because it can't happen.  I replaced the one in `onoff_request()`; please follow-up there.

mbolivar commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+ * notified when the operation completes.
+ *
+ * If the cancellation fails the service retains control of the client
+ * objecect, and the client must wait for operation completion.
```
```suggestion
 * object, and the client must wait for operation completion.
```

mbolivar reviewed (commented):
I've only looked at documentation.

Boy, the breathe package we use to pull doxygen into sphinx really doesn't work as desired sometimes.

pabigot commented:
DNM until #21584 is merged and the code updated to incorporate the checks for pre-kernel invocation.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+/**
+ * @brief Compile-time initialization of an on-off service.
+ */
+#define ONOFF_SERVICE_INITIALIZER(_start, _stop, _reset, _flags) { \
```
The intent here was to parallel what's done with `k_sem_init()` and `Z_SEM_INITIALIZER()` so people can implicitly initialize the service, if they know what they're doing.  I'll make this internal because explaining when and why you can and cannot do that is too complex.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+ * The reset function may be called only when
+ * `onoff_service_has_error()` returns true.
+ *
+ * Functions will be invoked from interrupt context only if the
```
No, that's a leftover because #21584 needs to get merged so `k_is_pre_kernel()` is documented and visible without having to include the system init header.  I'll revisit this when that's merged.

mbolivar replied:
>  I'll revisit this when that's merged.

That's been merged, so this is unblocked.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+ *
+ * These functions may be invoked from any context including
+ * pre-kernel, ISR, or cooperative or pre-emptible threads.
+ * Compatible functions must be isr-callable and non-suspendable.
```
Apparently I wasn't complete in my cleanup.  Since this is already blocked for another PR, I'll wait until I see whether the terminology is going to be acceptable before doing anything more.

mbolivar commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+ * @brief State associated with a client of an on-off service.
+ *
+ * Objects of this type are allocated by a client, which must use an
+ * initialization function (e.g. onoff_client_init_signal()) to
```
> Guarding the doxygen comment for onoff_client_init_signal() inside the `#if CONFIG_POLL` (which should be `#ifdef`) means that it doesn't show up in the rendered RST documentation, so this is a dangling reference in the generated docs.

Still a problem. Note the zephyr.doxyfile.in diff in my gist is not present in this PR, which means DOXYGEN is not defined.

pabigot replied:
Ugh.  I'm familiar with that technique and just assumed Zephyr had been using it without my knowledge.  Instead Zephyr appears to add `CONFIG_FOO=y` to `PREDEFINED` for over two dozen special cases.  Which is wrong in so many ways, including the fact that `CONFIG_FOO` is supposed to be `1` not `y`.  I think the minimal change is to add `CONFIG_POLL=y`, in a separate PR.

mbolivar replied:
Let's see if we can get the doxyfile patch merged instead: https://github.com/zephyrproject-rtos/zephyr/pull/21727

pabigot commented on doc/reference/kernel/other/resource_mgmt.rst:
```diff
@@ -0,0 +1,83 @@
+* The service transitions from ON to OFF when last client release is
+  received.
+* Each service configuration provides functions that implement the
+  transition from off to on, from on to off, and optionally from an
```
I was trying to distinguish a generic state from an inferred machine state.  Not worth the hassle; it's just "off" and "on" in the update.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+ * Note that the call to this function may succeed in a case where the
+ * actual request fails.  Always check the operation completion
+ * result.
+ *
```
@mbolivar here's the new attempt to describe the situation.

mbolivar replied:
Great example; that really cleared it up for me. Thanks.

dbkinder commented on doc/reference/kernel/other/resource_mgmt.rst:
```diff
@@ -0,0 +1,83 @@
+  operation is guaranteed to be initiated, but whether the operation
+  itself succeeds or fails is indicated through client notification.
+  The initiation functions can be invoked from pre-kernel, thread, or
+  ISR context, though contexts and states where the operation cannot
```
change to ``though in contexts...``

dbkinder commented on doc/reference/kernel/other/resource_mgmt.rst:
```diff
@@ -0,0 +1,83 @@
+  itself succeeds or fails is indicated through client notification.
+  The initiation functions can be invoked from pre-kernel, thread, or
+  ISR context, though contexts and states where the operation cannot
+  be started the function will result in an error.
```
add a comma: ``be started, the ...``

dbkinder commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+ * to on.  This must not be null.  Include @ref ONOFF_SERVICE_START_SLEEPS as
+ * appropriate in flags.
+ *
+ * @param stop the function used to (initiate a) transition from on to
```
got an extra ``to`` here

nashif commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,611 @@
+	};
+}
+
+#if defined(CONFIG_POLL) || defined(DOXYGEN)
```
the better solution is not to enclose APIs in #ifdefs and instead document that a certain API requires a Kconfig to be enabled, something like that...

```
 * @note
 *    @rst
 *    You should enable :option:`CONFIG_POLL` in your project
 *    configuration.
 *    @endrst
```

I am not a fan of adding this check for DOXYGEN. We are moving the logic from a simple list in doxyfile into the code which is a very gross workaround and might cause confusion and in some cases will not be as simple as here, depending on the logic.

mbolivar replied:
> We are moving the logic from a simple list in doxyfile into the code which is a very gross workaround and might cause confusion and in some cases will not be as simple as here, depending on the logic.

I don't agree that the doxyfile list is "simple". Adding CONFIG_FOO=y to it when we want to document some new API could result in unexpected changes elsewhere if the author expected some CONFIG_  defines to be unset when they placed their docstrings.

dbkinder reviewed (dismissed):
LGTM, with a couple of doc tweaks.

tyrantlizard77 reviewed (dismissed):
where are the z_vrfy_... functions for the system calls you added?
Your test case needs to exercise these system calls from user mode.

pabigot commented:
> where are the z_vrfy_... functions for the system calls you added?

Good question.  My original thought was these would be system calls since the onoff service structure might be owned by a device driver.

The closest analog to this API that I can see is the `device_pm_foo()` functions, which are not system calls, and are apparently intended to be invoked within the driver under some condition.

I don't see any way to do any validation of the arguments without making onoff service objects  K_OBJECTs, which at this time doesn't seem necessary.  And if there's no validation, I don't think they need to be syscalls.

So they're not anymore.  Instead where validation is required, e.g. requesting clocks, it can be done at the intermediate level, e.g. `clock_request()`, which can delegate to this API after validation.

mbolivar reviewed (changes requested):
I went through most of the test cases and have minor comments. I didn't look at the cancellation-related tests.

mbolivar commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,606 @@
+ *
+ * @note
+ *   @rst
+ *   This capability is available only when :option:`CONFIG_POLL` is
```
@nashif  with the DOXYGEN macro, calling this with `CONFIG_POLL` unset is a nice, fail-fast build error. Without it, it's accepted by the compiler but silently does nothing.

Do you really think that is better?

mbolivar replied:
Note that later changes made by @pabigot make this a runtime error rather a build-time error. So still errors out, but still not really as nice IMHO. I won't get in the way further however.

pabigot replied:
IIRC when I originally checked this I got a link-time error because the poll functions that the application used were not built in.  The changes for the runtime error cover a situation where the application doesn't require the poll functions, but still configures for a signal notification.

The goal of the later changes was to eliminate a path where there was a silent failure.

andrewboie commented:
> I don't see any way to do any validation of the arguments without making onoff service objects K_OBJECTs, which at this time doesn't seem necessary. And if there's no validation, I don't think they need to be syscalls.

OK, so if I understand you correctly, this API set is for use by device drivers and not directly by applications?

nordic-krch commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,606 @@
+ * @retval -EIO if service has recorded an an error
+ * @retval -EINVAL if the parameters are invalid
+ * @retval -EAGAIN if the reference count would overflow
+ * @retval -EWOULDBLOCK if the function was invoked from non-thread
```
there is another reason for that: if request was called during to_off transition and stop notification was called from interrupt context and start sleeps. I know that it will be quite complex for the user so ideally there should be provided a pattern how to handle it (in doc or dox). 

is that there right way of handling -EWOULDBLOCK?
```
err = onoff_request(srv,cli);
while (err == -EWOULDBLOCK && !(k_is_in_isr() || k_is_pre_kernel())) {
  /* if we are in thread context and service cannot be started it
   * means that we hit to off transition. Try again. */
    err = onoff_request(srv_cli);
}
if (err < 0) {
  return err;
}
```

pabigot replied:
> there is another reason for that [`-EWOULDBLOCK`]: if request was called during to_off transition and stop notification was called from interrupt context and start sleeps.

No, if `onoff_request()` is called while the service is being turned off then this call will return success.  The asynchronous operation result would be `-EWOULDBLOCK` if the off transition completes in non-thread context.

nordic-krch replied:
right, i confused return value with operation result which may be `-EWOULDBLOCK` in case when `onoff_stop_notify` is called from irq context, start sleeps and request came while stopping.

Still, it seems that fail safe handling is quite complex and it would be good to show patterns.

Here is something i came up with assuming that service has `ONOFF_SERVICE_START_SLEEPS` and `ONOFF_CLIENT_NOTIFY_SIGNAL` client.
```
bool reset_done;
do {
  err = onoff_request(srv, cli);
  if (err < 0) {
    break;
  }
  k_poll(..);
  k_poll_signal_check(cli->async.signal, &signaled, &result);

  if (result == 0) {
   break;
  } else if (result != -EWOULDBLOCK && !reset_done) {
     /* request failed, if reset is supported, attempt to reset? */
     err = onoff_service_reset(srv, cli);
      if (err < 0) {
        break;
      }
      k_poll(..);
      k_poll_signal_check(cli->async.signal, &signaled, &result);
      if (result < 0) {
        break;
      }
      reset_done = true; /* prevent forever loop if service is corrupted. */
  } else {
     /* WOULDBLOCK - corner case, start failed because attempted to start from wrong context*/
  }
} while (true); 
```

pabigot replied:
OK, depending on what that's meant to do it's not right in several ways.  I'll work up a sample to show how various situations need to be handled.

nordic-krch replied:
>  it's not right in several ways

that only proves that sample is needed 😄

pabigot replied:
Eh, maybe.  Some of the problems would be resolved by testing the code, but more generally the service reset API is present only as support for edge cases, where it is possible to recover from a service error.  I have yet to identify a use case where such recovery is possible, so I'd rather wait until we have something to work with rather than put a guess as to what it could look like in documentation and a sample.

https://gist.github.com/pabigot/6176d58fc60b6f3f870209368a6caae4 shows something that I think would work and documents the edge cases, though I haven't even compiled it.  If you want more than this I won't be able to get to it until the end of next week.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,522 @@
+		/* Reject if in a non-thread context and start could
+		 * wait.
+		 */
+		if ((k_is_in_isr() || k_is_pre_kernel())
```
this check is done twice. Would be good to extract to a function (`is_in_sleepable_context()`). What about context which has interrupts locked? See https://github.com/zephyrproject-rtos/zephyr/blob/8617af383d7173e04d2cce4739a687940faa7e16/subsys/logging/log_msg.c#L71

pabigot replied:
If anything gets extracted it should be `k_is_thread_context()` defined as `!(k_is_in_isr() || k_is_pre_kernel())`.  Which probably be a helper in `kernel.h`.

I don't think checking for locked interrupts is relevant.  AFAICT only the logging infrastructure is doing this, dunno why.

nordic-krch replied:
> I don't think checking for locked interrupts is relevant. 

Right, in case of logging we want to make sure that we block only in preemptable context. Even though, there is no kernel function, there could be local function that would encapsulate that.

pabigot replied:
It could be abstracted locally, but I don't see that as improving the code.  The current idiom should be good until `k_is_in_thread_context()` is warranted.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,522 @@
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) == 0) {
+		rv = -EINVAL;
```
what about having `-EALREADY` here? Since if user do according to the book and tries reset on failure then it indicates that someone else already performed reset.

pabigot commented:
Most uses will be associated with device drivers.  There are some situations where an application might make use of it directly, but I think anything that has to involve user mode can deal with validation through a wrapper.

pabigot commented on tests/lib/onoff/src/main.c:
```diff
@@ -0,0 +1,1164 @@
+	zassert_equal(rc, -EINVAL,
+		      "validate rel null srv");
+
+	rc = onoff_release(&srv, NULL);
```
changed message to match this and next instead of reordering.

pabigot commented on tests/lib/onoff/src/main.c:
```diff
@@ -0,0 +1,1164 @@
+
+	memset(&cli, 0xA3, sizeof(cli));
+	rc = onoff_request(&srv, &cli);
+	zassert_equal(rc, -EINVAL,
```
The one with 0xA3 tests for completely invalid content, e.g. nobody called init at all.

The one with 0x3C sets additional bits that cause a different branch of the validation to fail.

These came from looking at coverage testing and tweaking the content to take different branches.

mbolivar replied:
Since the numbers are magic and result from coverage testing, can you please add a comment with this information?

pabigot commented on tests/lib/onoff/src/main.c:
```diff
@@ -0,0 +1,1164 @@
+
+	zassert_equal(srv.refs, 0U,
+		      "reset req refs: %u", srv.refs);
+	zassert_false(onoff_service_has_error(&srv),
```
I've been bitten when changes during maintenance inadvertently corrupt the before state so the after check isn't verifying a change occurred.  So I'll often re-verify that the state is the intended initial state both before and after I invoke the operation that should cause the state to change.  The intervening check on signal IMO separates the previous check from the next operation by too much code.

pabigot commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,606 @@
+ * be notified.  The client will be notified through cli with an
+ * operation completion of `-ECANCELED`.
+ * @retval -EINVAL if the parameters are invalid.
+ * @retval -EWOULDBLOCK if cancellation was rejected because the
```
The user reaction must be to continue to wait for the result: the operation cannot be cancelled.  

I think the detail is warranted: the last client canceling is the only condition where this would happen: cancels requested while in-progress that are not the last one are processed immediately.

pabigot commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,522 @@
+		/* Stay on if release leaves a client. */
+		if (srv->refs > 1U) {
+			notify = true;
+			rv = 1;
```
Because I want to have `refs` updates done after all the logic is complete, rather than hidden inside a conditional in the switch statement.  This makes it parallel to the start logic.

mbolivar reviewed (approved):
I am good with this. My one remaining open comment is just a nice-to-have.

@nashif @dbkinder @pabigot would you consider this worthy of the release-notes label? (I think yes, personally.)

nordic-krch commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,603 @@
+	 * async infrastructure.  This field is the sole record of
+	 * success or failure for no-wait synchronous operations.
+	 */
+	int volatile result;
```
result is used only for spinwait mode when reading `onoff_client_fetch_result()`. Should be moved to async structure.

pabigot replied:
`onoff_client_fetch_result()` is not specific to spinwait.  It's useful (albeit not necessary) for callbacks, and useful and necessary for some use of signals as well.  For example, you can start multiple services that all notify through the same signal; the result stored in the signal will be the last one received, while the individual results can be obtained from the client structure.

nordic-krch replied:
> For example, you can start multiple services that all notify through the same signal

Can you do that? How you can synchronize when multiple source can set the signal? It might be that i just don't understand the use case. Anyway, if you see case for having it for other modes then i'm ok with it.

pabigot replied:
Sure; the client structure doesn't own the signal, it just uses one that the application provides.  It's easy to share a signal and when it's raised look at other state (here, the result in the client structure) to determine what's ready.  Saves memory to only require one signal structure.

nordic-krch commented on include/sys/onoff.h:
```diff
@@ -0,0 +1,603 @@
+	 * Callers must check for completions using
+	 * onoff_client_fetch_result().
+	 *
+	 * See async_operation_init_spinwait().
```
onoff_client_init_spinwait(). You could add @ref.

pabigot replied:
Fixed; thanks.

Adding `@ref` is not necessary and has no effect: writing it as a function call `foo()` would cause Doxygen to attempt to generate a link to the corresponding documented element.  That doesn't show up in Zephyr's generated documentation because it would require dropping into `@rst` mode and using a `:c:func:` prefix, which only works on whole paragraphs.

nordic-krch commented:
@dbkinder can you finish your review? I'm interested in having that in master as I would like to use it.

mbolivar commented:
> @dbkinder can you finish your review? I'm interested in having that in master as I would like to use it.

Unfortunately, he his no longer working on Zephyr.

nordic-krch commented:
@carlescufi can you merge it? I believe that it is well reviewed (i personally spent many many hours on that) and it's been out in review for some time.

carlescufi commented:
> @carlescufi can you merge it? I believe that it is well reviewed (i personally spent many many hours on that) and it's been out in review for some time.

Since this introduces a new public API, I will include it in tomorrow's API meeting and do one last pass there before we merge.

tbursztyka commented on lib/os/onoff.c:
```diff
@@ -0,0 +1,517 @@
+		return -EINVAL;
+	}
+
+	*srv = (struct onoff_service)ONOFF_SERVICE_INITIALIZER(start, stop, reset, flags);
```
Fix the 80 chars limit warning here

pabigot replied:
Done.

tbursztyka commented on tests/lib/onoff/src/main.c:
```diff
@@ -0,0 +1,1164 @@
+
+
+	/* Non-initial request from ISR is OK */
+	struct onoff_client isrcli;
```
I am surprised checkpatch did not raised a warning/error here. Never declare a variable in a middle of a code block

pabigot replied:
There's lots of precedent for it in Zephyr, and the feature was added twenty years ago in C99 which Zephyr implicitly requires.  Unless there's a documented coding standard that disallows this I'm going to keep it in place.

tbursztyka replied:
ok, after all even basic checkpatch warnings did not stop other PRs to get merged.

nordic-krch commented:
example implementation using `onoff` service is clock control driver for nRF platform:
https://github.com/nordic-krch/zephyr/commit/2d18fee3229e962137e84a79bceb7333d806124f

It simplifies the driver. Driver need to provide
- start function: https://github.com/nordic-krch/zephyr/blob/2d18fee3229e962137e84a79bceb7333d806124f/drivers/clock_control/nrf_power_clock.c#L63
- stop function https://github.com/nordic-krch/zephyr/blob/2d18fee3229e962137e84a79bceb7333d806124f/drivers/clock_control/nrf_power_clock.c#L73

onoff service structure is part of driver data. Branch contains also a draft of updated clock control API with request release calls (which are optional since some clocks may have only one user and having onoff service for that is an overkill).

carlescufi commented:
@tbursztyka and @andrewboie any objections to merging this?

tbursztyka commented:
@carlescufi no, though from PM point of view this is not solving it properly (i.e.: fully static device dependency tree, etc...) but I don't think this PR is really meant for it anyway since it only handles 2 states.
